### DISCUSSION TITLE
Specify path to license.lic at runtime

### DISCUSSION THREAD

**Discussion by genachka** [dashingsoft/pyarmor#1138] [Ideas]
Unless I missed something, the only ability to have `license.lic` not in the same directory is using symbolic link as described here: https://pyarmor.readthedocs.io/en/latest/advanced.html#storing-runtime-file-license-lic-to-any-location

This doesn't work in most cases for us. We distributed a binary made by pyinstaller for multiple operating systems and folks copy and move the binary around, with some Linux folks running it on a server that may have multiple people putting it in to their home folders. Often the Linux admin wants to put things in a shared folder, i.e. `/opt/acme/myapp` (both binary and `license.lic`) and when the user runs from their home `/home/userx` (i.e. `linuxbox1:/home/userx$ /opt/acme/myapp`) it will complain it can't find the license. The admin doesn't want to allow symbolic links for the users.

We'd like to be able to pass a command line switch to our app that can provide the path to where the `license.lic` file resides at runtime. Something like this:
`linuxbox1:/home/userx$ /opt/acme/myapp --license /opt/acme/license.lic --otherswitches`

we would parse the command line args within the app and when we see `--license` we'd pass to some pyarmor API the path where to find the `license.lic` file at runtime.

**Comment by jondy** [comment]
Environment variable `PYARMOR_LICENSE` could be used to tell obfuscated scripts where to find license.lic, for example
```
PYARMOR_LICENSE=/opt/acme/license.lic /opt/acme/myapp --otherswitches
```

**Comment by genachka** [comment]
@jondy oh? I missed that, thank you! 

We'd rather not make it obvious that PyArmor is used. Any way to change the name of that, setting a env name of our choosing which would also be consistent with our env naming conventions? Maybe as a parameter while obfuscating the scripts?

**Comment by jondy** [comment]
In next major version 8.0+, the license.lic path could be specified by obfuscation option.

**Comment by sevagas** [comment]
By obfuscation it would not be at runtime. I agree that at least it could be interesting to have a different name for PYARMOR_LICENSE. Or be able to set it's value at runtime for example in PyInstaller hook. 
A third good enough option would be that  PyArmor  check the executable folder and not only the running folder for the license file.

**Comment by sevagas** [comment]
It's really not good UX to force my users to set this environment variable if they want to call my software from a different path. I might need to revert to an old PyArmor version because of that.

**Comment by genachka** [comment]
@jondy Please elaborate on what that means. Meaning that I would have to know in advance during the obfuscation the path to specify? In which case that doesn't solve it as the idea is to allow user at runtime to control the location. Or do you mean that during obfuscation I could specify the environment variable name -- in which case it would exactly solve my need.

> In next major version 8.0+, the license.lic path could be specified by obfuscation option.

**Comment by jondy** [comment]
Actually for this case, there is a simple workaround and pyarmor need  do nothing.

Just add a few lines in the main script to copy `/path/to/license.lic` to pyarmor expected path

for example,  add code like this

```python
import pytransform
import shutil
licfile = '/path/to/license.lic'   # get real path from command line
dest = os.path.dirname(pytransform.__file__)
shutil.copy(licfile, dest)
```

**Comment by sevagas** [comment]
Thanks @jondy, thanks for that. In this case could it be possible to add this in the main script?
os.environ["PYARMOR_LICENSE"] = ".path\to\license.lic"

The thing is when I try to add this code in main script its too late I gate the error could not find module about main script.
Is there a location to place it?

**Comment by jondy** [comment]
This solution may not work, I forget one thing: The obfuscated scripts could not be executed if there is no valid license found.

**Comment by sevagas** [comment]
In my case I have added this in the pyinstaller runtime hook:

```
if isfile(join(dirname(sys.executable), 'license.lic')):
    if getcwd() != dirname(sys.executable):
        if not getenv("PYARMOR_LICENSE"):
            print("\n\n [!] You must set PYARMOR_LICENSE environment variable to point to your license when calling software from another directory. \n")
            print("  ex: set PYARMOR_LICENSE=%s \n" % join(dirname(sys.executable), 'license.lic'))
            sys.exit(0)
    with open(join(dirname(sys.executable), 'license.lic'), 'rb') as fs:
        with open(join(sys._MEIPASS, 'license.lic'), 'wb') as fd:
            fd.write(fs.read())
```
Setting environment variable in the runtime file does not work. What I don't understand is why I get the error if license.lic is in the same folder as the executable but I am running from a different folder.

**Comment by jondy** [comment]
The obfuscated scripts search `license.lic` by this way:
* first environment `PYARMOR_LICENSE`
* then package `pytransform`
* finally current path.

**Comment by sevagas** [comment]
Could it be possible to copy licence.lic to pytransform in the runtime hook like I did for sys._MEIPASS?
Note the code I used to copy to sys._MEIPASS is from a PyArmor documentation 
(https://github.com/dashingsoft/pyarmor/blob/master/docs/advanced.rst section Bundle Obfuscated Scripts To One Executable File). I don't see why we have to do this if license.lic is not searched in sys._MEIPASS

**Comment by jondy** [comment]
Refer to outer runtime key in Pyarmor 8.0+
